lib/spacekit-relay.js

```diff
@@ -0,0 +1,178 @@
+    }
+
+    this.tlsCertificateGenerator = new TlsCertificateGenerator();
+
+    this.backoff = Backoff.fibonacci({
+      randomisationFactor: 0.4,
+      initialDelay: 1000,
+      maxDelay: 5 * 60000
+    });
+
+    this.backoff.on('backoff', (number, delay) => {
+      log.debug(`Reconnecting in ${delay}ms...`);
+    });
+
+    this.backoff.on('ready', this.connect.bind(this));
+
+    this.connect();
+  }
+
+  connect () {
+    log.info(`SpaceKitRelay connecting to ${this.proxyUrl}...`);
+
+    this.ws = new WebSocket(this.proxyUrl, 'spacekit', {
+      headers: {
+        'x-spacekit-subdomain': this.config.relay,
+        'x-spacekit-username': this.config.username,
+        'x-spacekit-apikey': this.config.apiKey
+      }
+    });
+
+    this.ws.on('open', () => {
+      log.info(`Connected as ${this.hostname}.`);
+      this.backoff.reset();
+      this.tlsCertificateGenerator.generateCertificate(this.hostname);
+    });
+
+    let currentMessageHeader = null;
+
+    this.ws.on('message', (data) => {
+      if (!currentMessageHeader) {
+        currentMessageHeader = JSON.parse(data);
+      } else {
+        this.handleMessage(currentMessageHeader, data);
+        currentMessageHeader = null;
+      }
+    });
+
+    this.ws.on('close', () => {
+      log.warn('Lost connection to the service.');
+      this.backoff.backoff();
+    });
+
+    this.ws.on('error', (err) => {
+      log.error(err, 'Service connection encountered an error.');
+      this.backoff.backoff();
+    });
+  }
+
+  sendMessage (header, body) {
+    if (this.ws.readyState === WebSocket.OPEN) {
+      let _data = { header: header, bodyLength: body && body.length };
+      log.trace(_data, 'send message');
+
+      this.ws.send(JSON.stringify(header));
+      this.ws.send(body || new Buffer(0));
+    }
+  }
+
+  handleMessage (header, body) {
+    let id = header.connectionId;
+    let socket = this.outgoingSockets.get(id);
+
+    switch (header.type) {
+      case 'open':
+        if (this.tlsCertificateGenerator.hasValidCertificate()) {
+          // TODO: spacekit server should return 443, not 80
+          if (header.port === 80) {
+            header.port = 443;
+          }
+          let portInfo = this.config.portMap.get(header.port);
+          if (!portInfo) {
+            log.error(`no port mapping for ${header.port}`);
+            this.sendMessage({
+              connectionId: id,
+              type: 'close'
+            }, null);
+            break;
+          }
+          log.info(`Attempting to establish a connection to ${portInfo.hostname}:${portInfo.destinationPort}`);
+          // TODO: what if portInfo.hostname is equal to this.hostname? will that cause things to blow up?
+          socket = new TlsTerminatingSocket(
+            Net.connect(portInfo.destinationPort, portInfo.hostname),
+            this.tlsCertificateGenerator.createSecureContext());
+        } else {
+          // TODO: The SpaceKit server should explicitly state that this
```

Comment:
You mean non-acme request forwarding, I presume
On Tue, Apr 19, 2016 at 4:26 PM Reza Akhavan <notifications@github.com>
wrote:

> In lib/spacekit-relay.js
> <https://github.com/spacekit/spacekit/pull/33#discussion_r60329471>:
>
> > +          let portInfo = this.config.portMap.get(header.port);
> > +          if (!portInfo) {
> > +            log.error(`no port mapping for ${header.port}`);
> > +            this.sendMessage({
> > +              connectionId: id,
> > +              type: 'close'
> > +            }, null);
> > +            break;
> > +          }
> > +          log.info(`Attempting to establish a connection to ${portInfo.hostname}:${portInfo.destinationPort}`);
> > +          // TODO: what if portInfo.hostname is equal to this.hostname? will that cause things to blow up?
> > +          socket = new TlsTerminatingSocket(
> > +            Net.connect(portInfo.destinationPort, portInfo.hostname),
> > +            this.tlsCertificateGenerator.createSecureContext());
> > +        } else {
> > +          // TODO: The SpaceKit server should explicitly state that this
>
> I removed acme url request forwarding from the server
> spacekit/spacekit-service#8
> <https://github.com/spacekit/spacekit-service/pull/8>
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub
> <https://github.com/spacekit/spacekit/pull/33/files/da88d06eb7ef08d3bd77b6e0473b3d1c805f4627#r60329471>
>